jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -780,143 +781,128 @@ public static String parentPath(String p)
+                        return null;
+                    slash = true;
                     dots = 0;
                     break;
 
+                case '?':
+                case '#':
+                    // finish normalization at a query
+                    break loop;
+
                 case '.':
-                    switch (dots)
-                    {
-                        case 0:
-                            dots = 1;
-                            break;
-                        case 1:
-                            dots = 2;
-                            break;
-                        case 2:
-                            canonical.append("...");
-                            dots = -1;
-                            break;
-                        default:
-                            canonical.append('.');
-                    }
+                    // Count dots only if they are leading in the segment
+                    if (dots > 0)
+                        dots++;
+                    else if (slash)
+                        dots = 1;
+                    else
+                        canonical.append('.');
+                    slash = false;
                     break;
 
                 default:
-                    switch (dots)
-                    {
-                        case 1:
-                            canonical.append('.');
-                            break;
-                        case 2:
-                            canonical.append("..");
-                            break;
-                        default:
-                    }
+                    // Add leading dots to the path
+                    while (dots-- > 0)
+                        canonical.append('.');
                     canonical.append(c);
-                    dots = -1;
+                    dots = 0;
+                    slash = false;
             }
-
             i++;
         }
+
+        // process any remaining dots
+        if (doDots(canonical, dots))
+            return null;
+
+        // append any query
+        if (i < end)
+            canonical.append(path, i, end);
+
         return canonical.toString();
     }
 
     /**
-     * Convert a path to a cananonical form.
-     * <p>
-     * All instances of "." and ".." are factored out.
-     * </p>
+     * Convert a decoded URI path to a canonical form.
      * <p>
+     * All segments of "." and ".." are factored out.
      * Null is returned if the path tries to .. above its root.
      * </p>
      *
-     * @param path the path to convert (expects URI/URL form, encoded, and with path separators '/')
+     * @param path the decoded URI path to convert. Any special characters (e.g. '?', "#") are assumed to be part of
+     * the path segments.
      * @return the canonical path, or null if path traversal above root.
+     * @see #canonicalURI(String)
      */
-    public static String canonicalEncodedPath(String path)
+    public static String canonicalPath(String path)
```

Comment:
As above, it would be clearer to call this method `canonicalDecodedPath` - or to keep the api the same, add this method and make `canonicalPath` deprecated and refer to `canonicalDecodedPath`.

Reply:
These are different methods that do different things.  This is definitely a URI which is never decoded.
I will rename the argument and improve the javadoc

Reply:
I'm not keen to rename this method.  There is now only `canonicalPath` and `canonicalUri`, with javadoc making it clear what they are.